libs/contrib/Data/String/Parser/Expression.idr

```diff
@@ -0,0 +1,125 @@
+-- Port of https://hackage.haskell.org/package/parsec-3.1.13.0/docs/Text-Parsec-Expr.html
+-- Original License	BSD-style: https://hackage.haskell.org/package/parsec-3.1.13.0/src/LICENSE
+
+module Data.String.Parser.Expression
+
+import Control.Monad.Identity
+import Data.String.Parser
+
+public export
+data Assoc = AssocNone
+           | AssocLeft
+           | AssocRight
+
+public export
+data Operator a = Infix (Parser (a -> a -> a)) Assoc
+                | Prefix (Parser (a -> a))
+                | Postfix (Parser (a -> a))
+
+public export
+OperatorTable : Type -> Type
+OperatorTable a = List (List (Operator a))
+
+public export
+BinaryOperator : Type -> Type
+BinaryOperator a = List (Parser (a -> a -> a))
+
+public export
+UnaryOperator : Type -> Type
+UnaryOperator a = List (Parser (a -> a))
```

Comment:
Should these lists be non-empty? Cf. `Data.List1` in `base`.

Once you use `List1` the `toParserX` functions are essentially `choice . List1.toList`.

Reply:
That's a good idea, tried it out but I ran into some trouble getting it to play nice with the `foldr splitOp` stuff on line 96. Could require a more invasive change...

Reply:
Updated to use `List1`, and renamed some stuff. Not sure are the names much better but `ReturnType` was not very descriptive. 

The `toParserX` functions are now the same except for the signatures. Is there some type level magic I can use reduce that duplication? Tried a few things but no luck.